src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs

```diff
@@ -119,6 +121,503 @@ internal int EffectiveMaxConcurrencyLevel
         }
     }  // class ParallelOptions
 
+    #region Worker Bodies
+
+    internal interface IWorkerBody<in TValue>
+    {
+        void Body(TValue value);
+
+        virtual void Finally() { }
+    }
+
+    internal  interface IWorkerBodyWithIndex<in TValue, in TIndex> : IWorkerBody<TValue>
+        where TIndex: INumber<TIndex>
+    {
+        void SetIteration(TIndex index);
+    }
+
+    internal sealed class SimpleWorkerBody<TValue> : IWorkerBody<TValue>
+    {
+        private readonly Action<TValue> _body;
+
+        internal SimpleWorkerBody(Action<TValue> body)
+        {
+            _body = body;
+        }
+
+        public void Body(TValue index)
+        {
+            _body(index);
+        }
+    }
+
+    internal abstract class WorkerBodyWithIndex<TValue, TIndex> : IWorkerBodyWithIndex<TValue, TIndex>
```

Comment:
with little code duplication in the worker classes I can change them classes to structs and reduce allocations if that is desirable. 
Thoughts?